src/commissioner/Discoverer.cpp

```diff
@@ -0,0 +1,264 @@
+    ReturnErrorOnFailure(commissionee.OpenBle());
+    mBleConnection.ClearValue();
+    ChipLogProgress(Controller, "Commissionable node candidate identified via BLE");
+    return CHIP_NO_ERROR;
+}
+
+void BleDiscoverer::OnBleDiscoverySuccess(void * context, BLE_CONNECTION_OBJECT connection)
+{
+    BleDiscoverer * instance = static_cast<BleDiscoverer *>(context);
+    instance->TallyBleDiscovery(connection);
+}
+
+void BleDiscoverer::OnBleDiscoveryError(void * context, CHIP_ERROR err)
+{
+    ChipLogError(Controller, "BLE Commissionable Node Discovery failed: %s", ErrorStr(err));
+    BleDiscoverer * instance = static_cast<BleDiscoverer *>(context);
+    instance->TallyBleFailure();
+}
+
+#endif // CONFIG_NETWORK_LAYER_BLE
+
+#if CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+DnssdDiscoverer::DnssdDiscoverer(Controller::DeviceControllerSystemState & systemState, Delegate ** delegate) :
+    Joinable(delegate), mSystemState(systemState)
+{
+    mSystemState.Retain();
+}
+
+DnssdDiscoverer::~DnssdDiscoverer()
+{
+    mSystemState.Release();
+}
+
+void DnssdDiscoverer::Shutdown()
+{
+    if (InProgress())
+    {
+        mDNSResolver.Shutdown();
+        // TODO: Setting the instance delegate to nullptr is a workaround for
+        //       #13227.  If we do not do this, minimal mdns can call back into
+        //       our allocated delegate proxy after it is freed.
+        chip::Dnssd::Resolver::Instance().SetResolverDelegate(nullptr);
+        ReleaseShutdownToken();
+    }
+}
+
+CHIP_ERROR DnssdDiscoverer::StartDnssdDiscovery(SetupPayload & payload, Platform::SharedPtr<ShutdownToken> token)
+{
+    VerifyOrReturnError(token.get() != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    Dnssd::DiscoveryFilter filter;
+    filter.type = payload.isShortDiscriminator ? Dnssd::DiscoveryFilterType::kShortDiscriminator
+                                               : Dnssd::DiscoveryFilterType::kLongDiscriminator;
+    filter.code = payload.isShortDiscriminator ? (payload.discriminator >> 8) & 0xF : payload.discriminator;
+    ReturnErrorOnFailure(this->mDNSResolver.Init(mSystemState.UDPEndPointManager()));
+    this->mDNSResolver.SetResolverDelegate(this);
+    this->mDeviceDiscoveryDelegate = this;
+    ReturnErrorOnFailure(this->mDNSResolver.FindCommissionableNodes(filter));
+    RetainShutdownToken(token);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DnssdDiscoverer::GetNextDnssdCandidate(Commissionee & commissionee)
+{
+    ReturnErrorOnFailure(GetValidRecord());
+    ReturnErrorOnFailure(GetAddress(commissionee.mCommissionableNodeAddress, commissionee.mMrpConfig));
+#if CHIP_PROGRESS_LOGGING
+    char addressStr[Transport::PeerAddress::kMaxToStringSize];
+    commissionee.mCommissionableNodeAddress.Value().ToString(addressStr);
+    ChipLogProgress(Controller, "Commissionable node candidate identified at %s", addressStr);
+#endif // CHIP_PROGRESS_LOGGING
+    ++mIpIdx;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DnssdDiscoverer::GetValidRecord()
```

Comment:
nit: This function is not returning anything, as I understand it, is only modifying `mNodeIdx` and `mIpIdx` so it can be consumed elsewhere, it might be worth returning the pointer to the data structure or even index.

Reply:
> so it can be consumed elsewhere, it might be worth returning the pointer to the data structure or even index.

Also, this is fair and understandable.  But while discovery code in the sdk is usually conceived of as sharing some centralized resource, the design here is purposefully not-that, and is tuned to a use case where there is only one caller.

Here, we want a Discoverer instance that will do the right thing with respect to its particular, calling commissioning state machine instance, avoiding conflicts and interaction with other sdk activities, and then will clean itself up.  So even if index or pointer references were provided to callers (which would avoid the need to have side effects in these Next getters), what cannot be shared amongst instances is the Joinable shutdown logic, or even the browse logic.

This approach admits that these things cannot be shared, and so is tailored to most simply support the core function of serving one single commissioning state machine instance, offloading as much complexity from that caller as possible.

Reply:
I've added comments to make very clear how this is working.